### PR TITLE
[WIP] Migration from old codebase

### PR DESCRIPTION
**Pull request by thibaudcolas**
Mainly from @samanthaksanders so far.

### CODE REVIEW

File: quicktube.js [hunk 1]
```diff
@@ -1,283 +1,518 @@
-/**
- * Quicktube.js
- * http://springload.co.nz/
- *
- * Copyright 2015, Springload
- * Released under the MIT license.
- * http://www.opensource.org/licenses/mit-license.php
- */
```
Comment (thibaudcolas) [inline quicktube.js:8]: I've removed the copyright noticed and added a LICENSE file instead.

File: quicktube.js [hunk 2]
```diff
@@ -1,283 +1,518 @@
...
-    } else if (typeof module === 'object' && module.exports) {
-        // Node. Does not work with strict CommonJS, but
-        // only CommonJS-like enviroments that support module.exports,
-        // like Node.
-        module.exports = (root.quicktube = factory());
-    } else {
-        // Browser globals
-        root.quicktube = factory();
-    }
-}(typeof global !== 'undefined' ? global : this.window || this.global, function () {
```
Comment (thibaudcolas) [inline quicktube.js:24]: Removed the UMD module factory. Moving forward we only want to support CommonJS modules (`module.exports`) and ES6 modules (`export default`, not here ATM). 

See https://github.com/umdjs/umd for more info.

File: quicktube.js [hunk 3]
```diff
@@ -1,283 +1,518 @@
...
-            self.options = self.extend(self.options, options);
-            $("[data-quicktube-play]").on("click", function() {
-                self.onClick.call(self, $(this));
-            });
-            $("[data-quicktube-play]").on("keydown", function(e) {
-                if(e.keyCode == 13) {
-                    self.onClick.call(self, $(this));
+'use strict';
+
+const SUPPORTS_TRANSITIONS = 'transition' in document.body.style || 'webkitTransition' in document.body.style || 'MozTransition' in document.body.style || 'msTransition' in document.body.style || 'OTransition' in document.body.style;
```
Comment (thibaudcolas) [inline quicktube.js]: We should dump this – see http://caniuse.com/#feat=css-transitions.

File: quicktube.js [hunk 4]
```diff
@@ -1,283 +1,518 @@
...
+
+// TODO has youtube api improved since this was written in 2015?
+// Mobile Safari exhibits a number of documented bugs with the
+// youtube player API. User agent detection, but you'll live, my boy!
+// https://groups.google.com/forum/#!topic/youtube-api-gdata/vPgKhCu4Vng
+const isMobileSafari = () => (/Apple.*Mobile.*Safari/).test(navigator.userAgent);
```
Comment (thibaudcolas) [inline quicktube.js:7]: Good question.

File: quicktube.js [hunk 5]
```diff
@@ -1,283 +1,518 @@
...
+
+// Inject the YouTube API onto the page.
+// TODO It is bad to do this systematically outside of any `init` function –
+// - It should be done only on init.
+// - It should check whether `YT` exists already so it does not load it multiple times.
+const newScriptTag = document.createElement('script');
+newScriptTag.src = 'https://www.youtube.com/iframe_api';
```
Comment (thibaudcolas) [inline quicktube.js:14]: See comment in the code above.

File: quicktube.js [hunk 6]
```diff
@@ -1,283 +1,518 @@
...
+    activeClass: 'quicktube--playing',
+    pausedClass: 'quicktube--paused',
+    posterFrameHiddenClass: 'quicktube__poster--hidden',
+    supportsTransitions: SUPPORTS_TRANSITIONS,
+    setExplicitFrameHeight: false,
+
+    init(options) {
+        // this.options = this.extend(this.options, options);
+        // TODO is this what that extend function is actually trying to do?
+        this.options = Object.assign(this.options || {}, options);
```
Comment (thibaudcolas) [inline quicktube.js]: `Object.assign` overrides the values of the left-most object with the values of the right-most, so I think this should be:

```js
this.options = Object.assign({}, this.options, options);
```

File: quicktube.js [hunk 7]
```diff
@@ -1,283 +1,518 @@
...
+
+        const playButton = document.querySelector('[data-quicktube-play]');
+        const stopButton = document.querySelector('[data-quicktube-stop]');
+
+        playButton.addEventListener('click', this.onClick.bind(this, playButton), false);
+
+        playButton.addEventListener('keydown', (e) => {
+            if(e.keyCode == 13) {
```
Comment (thibaudcolas) [inline quicktube.js:54]: Should be a constant for the keyCode.

File: quicktube.js [hunk 8]
```diff
@@ -1,283 +1,518 @@
...
-                var obj = arguments[i];
-                if (!obj) {
-                    continue;
+            }
+        };
+
+        // listen for play, pause and end states
+        // also report % played every second
+        const onPlayerStateChange = (e) => {
+            e['data'] == YT.PlayerState.PLAYING && setTimeout(onPlayerPercent, 1000, e['target']);
```
Comment (thibaudcolas) [inline quicktube.js]: All of those `e['thing']` could be re-written as `e.thing`.

Comment (thibaudcolas) [reply quicktube.js]: Oh and this should be an `if` – not a weird inline `foo === bar && doThis()`.

File: quicktube.js [hunk 9]
```diff
@@ -1,283 +1,518 @@
...
+        }
 
-            if (!isMobileSafari()) {
-                if (this.jamesPlayer.playVideo) {
-                    this.jamesPlayer.playVideo();
-                }
-            }
+        // TODO does this ever get set to true? if so how?
+        // Doesn't appear to be an overridable setting
+        if (this.setExplicitFrameHeight) {
```
Comment (thibaudcolas) [inline quicktube.js]: I think you could do this:

```js
const Quicktube = require('quicktube');

Quicktube. setExplicitFrameHeight = true
// And then...
Quicktube.init();
```

I don't think that's a good idea though, so we should drop this.

File: quicktube.js [hunk 10]
```diff
@@ -1,283 +1,518 @@
...
 
-            if (self.setExplicitFrameHeight) {
-                $video.height($parent.outerHeight());
-            }
+        if (!parentEl.getAttribute('data-video-playing')) {
+            this.hidePosterFrame(poster);
+            this._players[parentId] = parentEl;
+            parentEl.classList.add(this.activeClass);
+            parentEl.classList.remove(this.pausedClass);
+            window.dispatchEvent(new Event('quicktube:play'));
```
Comment (thibaudcolas) [inline quicktube.js:191]: We should get rid of those custom events. Either don't do anything, or have a callback function given as an `option`.

File: quicktube.js [hunk 11]
```diff
@@ -1,283 +1,518 @@
...
+            return;
+        }
+
+        this.quicktubePlayer.pauseVideo();
+        playerEl.classList.remove(this.activeClass);
+        playerEl.classList.add(this.pausedClass);
+        this.showPosterFrame(playerEl.querySelector('[data-quicktube-poster]'));
+        playerEl.getAttribute('data-video-playing');
+        this._players[parentId] = false;
+        window.dispatchEvent(new Event('quicktube:pause'));
```
Comment (thibaudcolas) [inline quicktube.js:225]: Same comment about the events.

File: quicktube.js [hunk 12]
```diff
@@ -1,283 +1,518 @@
...
+    },
+
+    trackEvent(event) {
+        if (typeof window._gaq === 'object') {
+            window._gaq.push(['_trackEvent', event.eventCategory, event.eventAction, event.eventLabel]);
+        } else if (typeof window.ga === "function") {
+            window.ga('send', 'event', event.eventCategory, event.eventAction, event.eventLabel);
```
Comment (thibaudcolas) [inline quicktube.js]: Should remove the whole support from `_gaq` I think:

```js
 window.ga('send', 'event', event.eventCategory, event.eventAction, event.eventLabel);
```

File: quicktube.js [hunk 13]
```diff
@@ -1,283 +1,518 @@
...
+        } else {
+        }
+    },
+};
+
+// Export this to window directly.
+window.onYouTubeIframeAPIReady = () => {
+    Quicktube.init();
+};
```
Comment (thibaudcolas) [inline quicktube.js]: We should get rid of this – projects should require Quicktube and then initialise it directly with `init`.

**Review by thibaudcolas — COMMENTED** [review]
@samanthaksanders I've added some notes to help you with this.

File: quicktube.js (hunk 3 above)
Comment (samanthaksanders) [reply quicktube.js]: Have removed :)

File: quicktube.js (hunk 6 above)
Comment (samanthaksanders) [reply quicktube.js]: Done 👍

File: quicktube.js (hunk 9 above)
Comment (samanthaksanders) [reply quicktube.js]: Removed 🎉

File: quicktube.js (hunk 12 above)
Comment (samanthaksanders) [reply quicktube.js]: I trust you on this without explicitly knowing why. Have simply left the `window.ga === 'function'` condition

**Pull request closed without merging**